artichoke-backend/src/extn/core/time/args.rs

```diff
@@ -63,12 +63,47 @@ impl TryConvertMut<&mut [Value], Args> for Artichoke {
                     result.year = i32::try_from(arg).map_err(|_| ArgumentError::with_message("year out of range"))?;
                 }
                 1 => {
-                    // TODO: This should support 3 letter month names
-                    // as per the docs. https://ruby-doc.org/3.1.2/Time.html#method-c-new
-                    let arg = to_int(self, arg)?;
-                    let arg: i64 = arg.try_convert_into(self)?;
+                    // ```irb
+                    // 3.1.2 => Time.utc(2022, 2).month
+                    // => 2
+                    // 3.1.2 => class I; def to_int; 2; end; end
+                    // => :to_int
+                    // 3.1.2 => Time.utc(2022, I.new).month
+                    // => 2
+                    // 3.1.2 > Time.utc(2022, "feb").month
+                    // => 2
+                    // 3.1.2 > class A; def to_str; "feb"; end; end
+                    // => :to_str
+                    // 3.1.2 > Time.utc(2022, A.new).month
+                    // => 2
+                    // ```
+                    let month: i64 = if let Ok(month) = to_str(self, arg).and_then(|arg| {
+                        let month_str: &str = self.try_convert_mut(arg)?;
+                        match month_str {
+                            "jan" => Ok(1),
+                            "feb" => Ok(2),
+                            "mar" => Ok(3),
+                            "apr" => Ok(4),
+                            "may" => Ok(5),
+                            "jun" => Ok(6),
+                            "jul" => Ok(7),
+                            "aug" => Ok(8),
+                            "sep" => Ok(9),
+                            "oct" => Ok(10),
+                            "nov" => Ok(11),
+                            "dec" => Ok(12),
+                            // This error will never be seen due to the
+                            // fallback of to_int conversion below
+                            _ => Err(ArgumentError::with_message("unsupported mon value").into()),
+                        }
+                    }) {
+                        Ok(month)
+                    } else {
+                        let arg = to_int(self, arg)?;
+                        arg.try_convert_into(self)
+                    }?;
```

Comment:
I spent a little time puzzling over the nicest way to do this. It could be split up a little instead of chaining the `and_then` into the `if let`, however I think the logic still is "correct".

It's also very tempting to short circuit this with a type check on `Ruby::FixNum` since it's very likely the majority of the usage.